common/src/main/java/io/netty/util/internal/StringUtil.java

```diff
@@ -213,16 +240,17 @@ public static String toHexString(byte[] src, int offset, int length) {
     public static int decodeHexNibble(final char c) {
         // Character.digit() is not used here, as it addresses a larger
         // set of characters (both ASCII and full-width latin letters).
-        if (c >= '0' && c <= '9') {
-            return c - '0';
-        }
-        if (c >= 'A' && c <= 'F') {
-            return c - ('A' - 0xA);
-        }
-        if (c >= 'a' && c <= 'f') {
-            return c - ('a' - 0xA);
+        final int index = c;
+        if (index >= HEX2B.length) {
+            return -1;
+        }
```

Comment:
Idea for another variant to avoid the branch here, not sure if it would be better or worse though because prediction should be near-perfect anyhow: Extend the array to 256 elements and use `index & 0xff` instead of checking length. This should still catch all invalid chars since UTF16 guarantees high bit of the low byte is always set for non-ascii.

Reply:
I will check what happen on ASM with it :)
Tbh now I'm more of the idea of using the idea of Nitsan (64K table lookup), that would avoid using plPlatformDependent and is delivering the best performance by saving the bounds checks. 
Probably will put the decision under vote, because the trade-offs here are so different...

Reply:
> This should still catch all invalid chars since UTF16 guarantees high bit of the low byte is always set for non-ascii.

Cyril character `'с' 1089` seems to have 0 MSB in its low byte..?

Reply:
Thanks @trustin @franz1981 and sorry, I think my statement was just completely incorrect! :man_facepalming: Was mixing up properties of UTF-8 and UTF-16... should have given it more thought before suggesting. So please scratch that idea!